tests/test_provenance.py

```diff
@@ -124,3 +215,87 @@ def test_mutate_numpy(self):
         self.assertEqual(decoded.parameters.keep, False)
+
+
+class TestParseProvenance(unittest.TestCase):
+    def test_parse_provenance(self):
+        # Most testing of parsing is in the round tripping below
+        with self.assertRaises(ValueError):
+            ts = msprime.simulate(5)
+            prov = json.loads(ts.provenance(0).record)
+            prov['software']['name'] = 'skynet'
+            tables = ts.tables
+            tables.provenances.clear()
+            tables.provenances.add_row(json.dumps(prov))
+            command, prov = msprime.provenance.parse_provenance(tables.tree_sequence())
+
+
+class TestRoundTrip(unittest.TestCase):
+    """
+    Check that we can recreate tree sequences from their provenances
+    """
+
+    def verify_equal(self, ts1, ts2):
+        ts1_tables = ts1.dump_tables()
+        ts2_tables = ts2.dump_tables()
+        # We clear the provenances as they are not equal due to timestamps
+        ts1_tables.provenances.clear()
+        ts2_tables.provenances.clear()
+        self.assertEqual(ts1_tables, ts2_tables)
+
+    def test_simulate_round_trip(self):
+        pop_configs = [msprime.PopulationConfiguration(5) for _ in range(2)]
+        ts = msprime.simulate(population_configurations=pop_configs,
+                              migration_matrix=[[0, 1], [1, 0]],
+                              demographic_events=[msprime.SimulationModelChange(
+                                  time=lambda t: t+0.1),
+                              ])
+        command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
+        recreated = getattr(msprime, command)(**parsed_prov)
+        self.verify_equal(ts, recreated)
+
+        inds = np.array([1, 2, 3, 4, 5, 6])
+        parent_indices = np.array([
+            4, 5,
+            4, 5,
+            4, 5,
+            4, 5,
+            -1, -1,
+            -1, -1]).reshape(-1, 2)
+        times = np.array([0, 0, 0, 0, 1, 1])
+        is_sample = np.array([1, 1, 1, 1, 0, 0])
+        t = max(times)
+        model = msprime.WrightFisherPedigree()
+        ped = msprime.Pedigree(inds, parent_indices, times, is_sample,
+                               sex=None, ploidy=2)
+        ts = msprime.simulate(
+            sample_size=4,
+            pedigree=ped,
+            demographic_events=[
+                msprime.SimulationModelChange(
+                    t, msprime.DiscreteTimeWrightFisher(2))],
+            model=model)
+        command, parsed_prov = msprime.provenance.parse_provenance(ts, None)
+        recreated = getattr(msprime, command)(**parsed_prov)
+        self.verify_equal(ts, recreated)
+
+        ts = msprime.simulate(10, random_seed=5)
+        from_ts = tsutil.decapitate(ts, ts.num_edges // 2)
+        start_time = from_ts.tables.nodes.time.max()
+        final_ts = msprime.simulate(
+            from_ts=from_ts, start_time=start_time, random_seed=2)
+        command, parsed_prov = msprime.provenance.parse_provenance(final_ts, None)
```

Comment:
No point in specifying the provenance=None here, is there?

Reply:
Yeah, that line is gone now after factoring out `verify`.